### DISCUSSION TITLE
Using TX (Gpio-01) and RX (Gpio-03) in tasmota device to connect Relay and a Switch.

### DISCUSSION THREAD

**Discussion by NsinghP** [arendst/Tasmota#12250] [Q&A] [answered]
In the ESP8266-12F based custom board I have designed I can use all GPIO-s as I wish. I am using GPIO-01 (TX) to connect a Relay and GPIO-03 (RX) to connect a switch.  I have already noticed that when I flash my custom .bin file to this device I lose the ability to transmit or receive any serial data from the device after successful flash. (For example "Send Config" and "Get IP" option in Tasmotizer1.2 stops working). Otherwise the device seems to be working as desired. My question is will, or can, this setup also effect normal working of Tasmota software in long term routine usage? Does the Tasmota firmware or ESP8266-12F itself need to use RX or TX Pins during routine operation without any external serial device connected to Rx and Tx pins? Has anyone experienced any problems after using RX or TX pin or both pins as in this case? What Kind of problems can i expect in routine operation of this device? Everyone with any Idea is requested to share their insight/experience.

Regards,

Navi
![Custom Board](https://user-images.githubusercontent.com/69470381/120271783-8b869480-c2c9-11eb-8875-d36f76d4759f.jpeg)

**Comment by barbudor** [comment]
Tasmota will not start to control those pins until a template that enforce them as Relay and Button/Switch is activated. So until then the pins could be able to keep their original usage if te hardware doesn't prevent it.
On RX, if the hardware is just a pull-up and a push-to-make to the ground, réception should work. But if you have the button/switch forcing down the pin, then it will not work. 
On TX it may depends on your circuit to drive the relay. A classical serie resistor to the base of a NPN transistor should be ok is the resistor do not  too small.
Best would be to use an oscilloscope to see what is happening (or sharing your circuit for better advice) 

Note that if the template is hard-coded in the binary, like on your other discussion, you're done. Tasmota will take control of the pins from the beginning on the 1st boot.

You can check by running Termite/Putty or other serial terminal after flashing and try basic commands.

**Comment by sfromis** [comment, marked as answer]
As TX and RX (along with gpio 2 and 16) from ESP8266 are high at boot, they are not ideal for connecting a relay, which you may want better control over. If your setup can somehow pull TX down, boot will fail. Whenever using tx/rx differently, I prefer to ensure having `SerialLog 0` to dissociate Tasmota from serial activity. In general, there is no harm in taking over tx/rx for other purposes, as long as you're fine with the limitations.

**Reply by barbudor** [reply]
Even with Tasmota's `SerialLog 0` you can't prevent serial activity at boot from ESP core. So your relay will buzz (at 115000 bps) during first 5 seconds until Tasmota takes control.

**Reply by NsinghP** [reply]
> 
> 
> Even with Tasmota's `SerialLog 0` you can't prevent serial activity at boot from ESP core. So your relay will buzz (at 115000 bps) during first 5 seconds until Tasmota takes control.

So I am assuming this activity will happen on TX. But so far in my testing the relay seems to be doing allright. I don't notice any unusual side effects. What I am worried about is connected light or fan suddenly turning on or off. (momentary on then off or momentary off than on is acceptable as long as it goes back to previous state). I have same worry about Rx line where the switch is connected--what if a serial disturbance changes the state of switch? is that possible?

**Reply by NsinghP** [reply]
> Whenever using tx/rx differently, I prefer to ensure having `SerialLog 0` to dissociate Tasmota from serial activity.

so what are the side effects of Seriallog 0? will tasmota web-console still work?

**Reply by barbudor** [reply]
Yes, web will still work
Only serial console will be disabled

**Reply by sfromis** [reply]
OTOH, sometimes there are side effects of *not* using `SerialLog 0` when you do not want a serial console.

However, may main point was not about serial activity, but that you should expect TX and RX to be briefly high at power on, which may not be so great for the purpose of your relays.

**Reply by NsinghP** [reply]
> 
> 
> OTOH, sometimes there are side effects of _not_ using `SerialLog 0` when you do not want a serial console.
> 
> However, may main point was not about serial activity, but that you should expect TX and RX to be briefly high at power on, which may not be so great for the purpose of your relays.

Good to know! Though I don't have full understanding, but in my setup so far, I don't see any undesirable behavior at boot. I am still going to set seriallog to 0, but first I need to figure out how to set seriallog to 0 in the custom .bin file so that serial log is set to 0 by default.

EDIT: I think #define SERIAL_LOG_LEVEL       LOG_LEVEL_NONE    // Should do the trick

**Reply by NsinghP** [reply]
In my latest flashed device  "Configure Module" page shows:
![Configure Module](https://user-images.githubusercontent.com/69470381/120845812-ef8daf00-c58e-11eb-878a-a9ec7f00996e.jpeg)
and "Configure Template" page shows:
![Configure Template](https://user-images.githubusercontent.com/69470381/120845886-06340600-c58f-11eb-829e-40c8b8cda84e.jpeg)

I am assuming that "Configure Module" Page only shows the unconfigured GPIO-s.

My Questions are: 

(1) Should I configure these remaining GPIOs to lets say GPIO15 to TX and GPIO16 to RX? Not that I need to do that but was just curious if i ever wanted to connect a serial logging interface to my already flashed device.

(2) I read that GPIO16 needs to be connected to RESET to activate "Deep Sleep" does this have any implication on "Setoption60" feature of Tasmota Firmware or does "Setoption60" works regardless of connecting GPIO16 to reset? In my case GPIO16 is not connected to reset yet.

(3) Since I am only using this device to turn on/off lights and fans, should I even care about using SetOption60 0 and Sleep xx and If I should then what value should I assign to Sleep xx? Should I use recommended optimal value of 75?

Regards,

Navi

**Reply by barbudor** [reply]
1) Generally what is the Template is what match the hardware and available/reachable pins are set to "User" so that they appear in "Configure Module". Pins that are not usable should be set to "Noe".
For example on a Sonoff Basic, pins for Relay, button, Leds are defined by the hardware and would be defined in the Templates. Some pins are available for the "User" customization and some pins are "None" available.
A mini-D1 would have all the Template pins set to User and their usage define in "Configure Module".
At the end, you do what you want.

Tasmota SerialLog is available ONLY on standard TX and RX and should be left unused (None or User in both pages).
If you assign a function to those pins, even as Serial, you should disable SerialLog.
You can define other pins to SSerialTX/RX to use commands SSend<x>. That will use software serial.

2) Dynamic Sleep as nothing to do with DeepSleep
DeepSleep stops the CPU totally and configure an hardware timer that will trigger GPIO16. Connecting GPIO16 to RESET will restart the CPU. This is used for battery operated device.
If you are using DeepSleep(Time) the CPU will stops entering and DeepSleep mode, and if GPIO16 is not connect to RESET it will remains stopped until you manual reset or power cycle it.

3) `SetOption60 0` is default. Read the article about DynamicSleep to understand if you have a use case that would need NormalSleep

**Reply by sfromis** [reply]
Better let unused pins stay as "User" if available for future projects, or "None" if not practically usable. Unless it is ESP8285, you should *always* have gpio9 and 10 as "None", as these are never available on ESP8266 running Tasmota.

**Reply by NsinghP** [reply]
Thanks, let me read up more on this. From your reply seems like I don't need to worry about connecting GPIO16 to RST, since my device is not battery operated  or worry about assigning new RX/TX pins. However, what if even without a battery I was interested in power saving benefits of deepsleep? is it worth messing around with? I read the article on DynamicSleep and feel that since I have no sensitive sensors that need monitoring I should go for a large Sleep XX value, maybe going as high as maximum 250!

**Reply by sfromis** [reply]
When having pins configured for buttons/switches, deep sleep does not make much sense if they are to be usable. The device can neither be woken up by signals on those pins, or via remote commands to operate relays.

**Reply by NsinghP** [reply]
> 
> 
> When having pins configured for buttons/switches, deep sleep does not make much sense if they are to be usable. The device can neither be woken up by signals on those pins, or via remote commands to operate relays.

That rules out 'DeepSleep" circuit. Good to know, Thanks.
